projects/extension/src/background/index.ts

```diff
@@ -192,60 +97,107 @@ chrome.runtime.onMessage.addListener(
+            tab: {
+              id: sender.tab!.id!,
+              url: sender.tab!.url!,
+            },
           })
 
-        chains.get(sender.tab!.id!)!.chains.set(message.chainId, {
-          chainName: message.chainSpecChainName,
-          peers: 0,
+          environment.set({ type: "activeChains" }, chains)
+          notifyChainsChanged()
         })
-        notifyAllChainsChangedListeners()
         break
       }
 
       case "chain-info-update": {
-        const info = chains.get(sender.tab!.id!)!.chains.get(message.chainId)!
-        info.peers = message.peers
-        info.bestBlockNumber = message.bestBlockNumber
-        notifyAllChainsChangedListeners()
+        environment.get({ type: "activeChains" }).then((chains) => {
+          if (!chains) return
+
+          const pos = chains.findIndex(
+            (c) =>
+              c.tab.id === sender.tab!.id! && c.chainId === message.chainId,
+          )
+          if (pos !== -1) {
+            chains[pos].peers = message.peers
+            chains[pos].bestBlockHeight = message.bestBlockNumber
+          }
+
+          environment.set({ type: "activeChains" }, chains)
+          notifyChainsChanged()
+        })
         break
       }
 
       case "database-content": {
-        chrome.storage.local.set({
-          [message.chainName]: message.databaseContent,
-        })
+        environment.set(
+          { type: "database", chainName: message.chainName },
+          message.databaseContent,
+        )
         break
       }
 
       case "remove-chain": {
-        chains.get(sender.tab!.id!)!.chains.delete(message.chainId)
-        notifyAllChainsChangedListeners()
+        environment.get({ type: "activeChains" }).then((chains) => {
+          if (!chains) return
+          const pos = chains.findIndex(
+            (c) =>
+              c.tab.id === sender.tab!.id! && c.chainId === message.chainId,
+          )
+          if (pos !== -1) chains.splice(pos, 1)
+          environment.set({ type: "activeChains" }, chains)
+          notifyChainsChanged()
+        })
         break
       }
     }
   },
 )
 
 chrome.tabs.onRemoved.addListener((tabId) => {
-  chains.delete(tabId)
+  environment.get({ type: "activeChains" }).then((chains) => {
+    if (!chains) return
+
+    while (true) {
+      const pos = chains.findIndex((c) => c.tab.id === tabId)
+      if (pos === -1) break
+      chains.splice(pos, 1)
+    }
+
+    environment.set({ type: "activeChains" }, chains)
+    notifyChainsChanged()
+  })
 })
 
-chrome.storage.local.get(["notifications"], (result) => {
-  if (Object.keys(result).length === 0) {
-    // Setup default settings
-    chrome.storage.local.set({ notifications: settings.notifications }, () => {
-      if (chrome.runtime.lastError) {
-        console.error(chrome.runtime.lastError)
-      }
-    })
-  }
+// TODO: right now it's ok, but will be wrong with manifest v3, because the script might reload
+// TODO: ?!?! why do we need to do this?
```

Comment:
This is for the `notifications` setting - were user can receive popup notifications based on alerts provided by the extension. It was erroneously removed during a previous refactor and it is planned to be added in the `settings` section